File: mixed/f1/group2/BinaryTreeZigzagLevelOrderTraversal.java
// Given the root of a binary tree, return the zigzag level order traversal of its nodes'
// values. (i.e., from left to right, then right to left for the next level and alternate between).
// Sample 1:
// Input: root = [3,9,20,null,null,15,7]
// Output: [[3],[20,9],[15,7]]
// Note: visualize to understand
// It is row level traversal starting from first row, which is root node. First left to right, then right to
// left and again left to right and so on.


import java.util.ArrayList;
import java.util.List;
import java.util.LinkedList;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class BinaryTreeLevelOrderTraversal {
    public static void main(String...args) {
        TreeNode tree = new TreeNode(3);
        TreeNode left = new TreeNode(9);
        TreeNode right = new TreeNode(20, new TreeNode(15), new TreeNode(7));
        tree.left = left;
        tree.right = right;

        List<List<Integer>> zOrder = zigzagLevelOrder(tree);
        for (List<Integer> elem: zOrder) {
            System.out.print(elem + ", ");
        }
        System.out.println();
    }

    static List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> sol = new ArrayList<>();
        traverse(root, sol, 0);

        return sol;
    }

    static void traverse(TreeNode curr, List<List<Integer>> sol, int level) {
        if (curr == null)
            return;

        if (sol.size() <= level) {
            List<Integer> newLevel = new LinkedList<>();
            sol.add(newLevel);
        }

        List<Integer> collection = sol.get(level);
        if (level % 2 == 0)
            collection.add(curr.val);
        else collection.add(0, curr.val);

        traverse(curr.left, sol, level + 1);
        traverse(curr.right, sol, level + 1);
    }
}

File: bytopic/binarytreebfs/python/group2/binary_tree_zigzag_level_order_traversal.py
"""
Given the root of a binary tree, return the zigzag level order traversal of its nodes'
values. (i.e., from left to right, then right to left for the next level and alternate between).

Sample 1:
Input: root = [3,9,20,null,null,15,7]
Output: [[3],[20,9],[15,7]]
Note: visualize to understand
It is row level traversal starting from first row, which is root node. First left to right, then right to
left and again left to right and so on.

Tag: 85/150
Tag: 103/2927, R697/2936 (overall frequency ranking)
"""
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def zigzag_level_order(root: Optional[TreeNode]) -> list[list[int]]:
    if not root:
        return[]
    res = []
    queue = deque([root])
    even_level = False
    while queue:
        n = len(queue)
        level = deque()
        for _ in range(n):
            node = queue.popleft()
            if even_level:
                level.appendleft(node.val)
            else:
                level.append(node.val)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        res.append(list(level))
        even_level = not even_level
    return res


def main():
    root = TreeNode(3, TreeNode(9), TreeNode(20))
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)

    print(zigzag_level_order(root))


if __name__ == "__main__":
    main()
